### PR TITLE
feat: add support for 3ds and surcharge decision through routing rules

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

added support for deriving 3ds and surcharge decisions through routing rules set by merchant.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Steps to test : https://github.com/juspay/hyperswitch-cloud/issues/2731#issuecomment-1770147086

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by Narayanbhat166 — DISMISSED** [review]

File: crates/router/src/core/payments.rs [hunk 1]
```diff
@@ -36,11 +38,17 @@ pub use self::operations::{
...
     flows::{ConstructFlowSpecificData, Feature},
+    helpers::get_key_params_for_surcharge_details,
     operations::{payment_complete_authorize, BoxedOperation, Operation},
     routing::{self as self_routing, SessionFlowRoutingInput},
 };
-use super::errors::StorageErrorExt;
+use super::{
+    errors::StorageErrorExt,
+    payment_methods::surcharge_decision_configs::perform_surcharge_decision_management_for_session_flow,
+    utils::persist_individual_surcharge_details_in_redis,
```
Comment (jarnura) [inline crates/router/src/core/payments.rs:50]: Remove here and add self in utils and use utils::fn directly, for payment_method also follow the same method.

**Review by jarnura — COMMENTED** [review]
Add instruments to the function which does external calls like database and redis,etc.

Add instruments to the functions which invoked in payments core flows.

**Review by Aprabhat19 — DISMISSED** [review]
I checked the 3ds flow , its working as expected.

**Comment by hrithikesh026** [comment]
> Overall LGTM. @hrithikesh026 Please test the surcharge flow in this PR once @Aprabhat19 Please test the 3DS decision manager flow in this PR once.

Tested surcharge feature and found out the surcharge was getting updated in DB from request. Fixed it. Everything else works fine.

**Review by Narayanbhat166 — APPROVED** [review]

**Review by vspecky — APPROVED** [review]

**Pull request merged**